### ISSUE TITLE
Implement moving in straight lines

### ISSUE THREAD

**Issue by SafwatHalaby** [cuberite/cuberite#2772]
Mob always move snapped to the grid. They should ditch the pathfinder and move straight towards the player when both of these condition are met:
 1. The mob won't collide with anything if moving in a straight line. (Simple line tracing is not enough, since mobs have varying bounding boxes)
 2. Both the mob and the target are at the same height, and they won't encounter any bumps or holes when moving in a straight line towards target.

*Performance considerations:*
 - This would call A* less than before. Performance should increase.
 - Don't do these checks every single tick.

*Gameplay benefits:*
 - Mobs will move more realistically and won't insist on staying grid-aligned.

**Comment by madmaxoft** [comment]
How about instead applying this kind of post-processing to each Y-level of the A*-ed path? This would make mobs walk naturally everywhere and could be easier to implement rather than the "special cases" scenario.

**Comment by SafwatHalaby** [comment]
That's a good point to consider. Applying this as A* post-processing would be far more naturally looking. I doubt it would be easier to implement though. I think performance would be worse.

**Comment by worktycho** [comment]
Straight line routes should be trivial to post process. Routes with bends would be expensive, so we only want to do that if we could work out some way of caching the paths. I can't think of a good way of handling block updates though.

**Comment by SafwatHalaby** [comment]
Block updates are already somehow handled: The mob would bump into newly placed blocks, give up and recalc.

**Comment by worktycho** [comment]
Only for paths. If we added another layer of caching containing additional connections, several mobs would bump into the block before solving.

But if thats acceptable we could have a process as part of the path finding to look at if any two node path can be traveled directly. This would be a relatively cheap way to postprocess. we would then have a map of point to point paths that would cache the routes that could be traveled. Over time the system would build up long paths in places that don't change frequently. The map would have to expire nodes if they aren't used to avoid growing too large. You would also need to split the cache by bounding box size. Then when a path is invalidated due to being impassable for some reason, you remove the edge currently traversed from the cache.

**Comment by SphinxC0re** [comment]
That sounds like you want to implement a neuronal network for pathfinding. Awseome!

**Comment by worktycho** [comment]
Whats a neuronal network? Google thinks I want neural network, which is very different from what I want to do here.

**Comment by worktycho** [comment]
I'd say this is more like a combination of a cellular automata, and graphs.

**Comment by SafwatHalaby** [comment]
@worktycho You remind me of some other method I considered earlier: Instead of A*, implement Jump Point Search and cache the Jump points.

**Comment by SafwatHalaby** [comment]
That would be heavy initially, but after the map is built pathfinding is essentially free.

**Comment by SafwatHalaby** [comment]
Actually, why would it be heavy initially? I think it wouldn't.

**Comment by worktycho** [comment]
The algorithm would be interesting to combine with jump point search, because the less nodes on the path, the faster it will run. Its key advantage over jump point search is that it will eventually converge to include the shortest euclidean route.

**Comment by worktycho** [comment]
And it shouldn't be heavy initially, because it runs in O(N) after the path is found.

**Comment by SafwatHalaby** [comment]
>Its key advantage over jump point search is that it will eventually converge to include the shortest euclidean route.

I didn't get that part.

Note that JPS rids you of the postprocessing part, you get a ready-to cache Jump Points.

**Comment by SafwatHalaby** [comment]
Note that one of them problems with cache (Which, among other things, made me back away from the JPS idea) is that mobs with different bounding boxes would need different caches. (E.g. a spider can walk in a narrow tunnel but a zombie can't)

**Comment by worktycho** [comment]
Yes. But mobs could use edges from caches with larger boxes. This might result in most of the edges converging into the caches for the largest boxes.

**Comment by SafwatHalaby** [comment]
Sounds overengineered to me.

**Comment by tigerw** [comment]
This entire thing sounds over engineered; I think novices like me would greatly appreciate if the pathfinder wasn't turned into something like either:

1. The current redstone simulator or
2. The world generation code

I think that all you need is to get the path generation algorithm to do diagonals (rather than right angled bends) if it finds nothing in the way, and assign these to a lower cost. Is that how things work?

**Comment by worktycho** [comment]
You can get it to do 45 degree bends relatively easily, but on large flat area that can still lead to weird looking paths. Doing if you want the ability to do straight paths between any arbitrary points on a plane, you need complexity or horribly slow paths. The naive algorithm is something like O(N^4).

**Comment by SafwatHalaby** [comment]
It already does 45 degree bends. And it looks weird often.
The original suggestion is not over engineered, and is probably O(N) with some pretty small constant.
```
cPathFinder::getNextWayPoint()
{
    if (I can move in straight line to the player)
    {
       move in straight line
    }
    else
    {
       Summon cPath.
    }
}
```
